nat-lab/tests/test_dns.py

```diff
@@ -632,48 +639,30 @@ async def test_dns_duplicate_requests_on_multiple_forward_servers() -> None:
                 SetupParameters(
                     connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_1,
                     ip_stack=IPStack.IPv4v6,
-                    connection_tracker_config=generate_connection_tracker_config(
-                        ConnectionTag.DOCKER_CONE_CLIENT_1
-                    ),
+                    connection_tracker_config=[
+                        ConnectionTrackerConfig(
+                            key="dns-limiter",
+                            limits=ConnectionLimits(
+                                1, 1
+                            ),  # Require strictly one DNS connection
+                            target=FiveTuple(
+                                protocol="udp",
+                                src_ip=LAN_ADDR_MAP[ConnectionTag.DOCKER_CONE_CLIENT_1],
```

Comment:
Now, the test will succeed if we don't use magic dns but contact any of the external dns servers. Please add something like `dst_ip=FIRST_DNS_SERVER`, and possible make the list of configured magic dns servers have one element. If we go with reduction of the list down to one entry, a new test will have to be added to verify that configuring longer list of IPs as DNS servers works as expected.

Alternatively, maybe we could extend ConnectionTracker to expect no only literal value but also predicates for the `FiveTuple` attributes?

Reply:
Well we should not restrict DNS server to choose first server for first query necessarily, generally those two servers should be treated equally. 

I guess one would have to implement predicates in order to allow for such flexibility, but I wonder if it is not overkill for this test. On the other hand, I guess predicates could prove useful in general case, not just for this test.

Reply:
We can reach any outside dns server from our nat-lab network, I can easily imagine someone hardcoding 8.8.8.8 (mistake as part of some future feature work). This kind of mistake should cause this test to fail, but right now it will not.

Reply:
Refactored